clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -75,6 +82,8 @@
 
     /* max time in ms for the producer to wait for acknowledgement from server*/
     private final int requestTimeoutMs;
+    
+    private final Time time;
```

Comment:
It seems a little unfortunate to use Time internally while still supporting all the methods which set the timestamp directly. Since the only call where we actually use this reference is poll(), perhaps we could drop the local reference and change the signature of poll() to:
```java
public List<ClientResponse> poll(long timeout, Time time);
```
This is still very-low impact and would solve the issue KAFKA-2615. Of course, the bigger change would be to allow the local reference and to refactor the other methods to avoid sending the timestamp. If we could figure out a way to do that without requiring excessive calls to time.milliseconds(), I think it might be preferable since it would clean up the API and possibly avoid the need to pass Time around everywhere.